Summarize Java compilation failures

At the moment, when javac fails we only report the generic Gradle error ("Compilation failed; see the compiler error output for details."). We should grab the compiler output and provide a better BLUF.

Current plan of attack:
 - [ ] Grab standard error (`compileJava.listener.addStandardErrorListener`)
 - [ ] Parse standard error (look for `error:` and grab that line + following context lines)
 - [ ] Dump into a junit report
 - [ ] If parsing failed to find an error but the task failed, fall back to something sensible (e.g. dumping the whole of stderr into the junit report)

Example standard error:
```
> Task :javaCompileCircleFinalizer
warning: [options] bootstrap class path not set in conjunction with -source 1.7
/private/var/.../src/main/java/com/example/MyClass.java:8: error: incompatible types: String cannot be converted to int
    private final int a = "hello";                               
                          ^
/private/var/.../src/main/java/com/example/MyClass.java:12: error: cannot assign a value to final variable b
        b = 2;                                                   
        ^                                                        
Note: /private/var/.../src/main/java/com/example/MyClass.java uses unchecked or unsafe operations.
Note: Recompile with -Xlint:unchecked for details.               
2 errors                                                                                                                           
1 warning
```

--- a/src/main/java/com/palantir/gradle/circlestyle/CircleStylePlugin.java
+++ b/src/main/java/com/palantir/gradle/circlestyle/CircleStylePlugin.java
@@ -24,6 +24,7 @@ import org.gradle.api.Plugin;
 import org.gradle.api.Project;
 import org.gradle.api.plugins.quality.Checkstyle;
 import org.gradle.api.plugins.quality.FindBugs;
+import org.gradle.api.tasks.compile.JavaCompile;
 
 public class CircleStylePlugin implements Plugin<Project> {
 
@@ -60,6 +61,16 @@ public class CircleStylePlugin implements Plugin<Project> {
                                 timer,
                                 XmlReportFailuresSupplier.create(findbugs, new FindBugsReportHandler()),
                                 new File(circleReportsDir, "findbugs"));
+                    }
+                });
+                project.getTasks().withType(JavaCompile.class, new Action<JavaCompile>() {
+                    @Override
+                    public void execute(JavaCompile javac) {
+                        registerFinalizer(
+                                javac,
+                                timer,
+                                JavacFailuresSupplier.create(javac),
+                                new File(circleReportsDir, "javac"));
                     }
                 });
             }

--- a/src/main/java/com/palantir/gradle/circlestyle/Failure.java
+++ b/src/main/java/com/palantir/gradle/circlestyle/Failure.java
@@ -27,6 +27,12 @@ interface Failure {
     int line();
     String severity();
     String message();
+    String details();
 
-    class Builder extends Failure_Builder { }
+    class Builder extends Failure_Builder {
+        public Builder() {
+            source("");
+            details("");
+        }
+    }
 }

--- a/src/main/java/com/palantir/gradle/circlestyle/FailuresReportGenerator.java
+++ b/src/main/java/com/palantir/gradle/circlestyle/FailuresReportGenerator.java
@@ -36,16 +36,16 @@ public class FailuresReportGenerator {
                 .subname(taskName);
 
         for (Failure failure : failures) {
-            String shortSource = failure.source().replaceAll(".*\\.", "");
+            String shortSource = failure.source().isEmpty() ? "" : failure.source().replaceAll(".*\\.", "") + " - ";
             String className = getClassName(failure.file());
 
             Report.TestCase testCase = new Report.TestCase.Builder()
-                    .name(shortSource + " - " + className)
+                    .name(shortSource + className)
                     .failure(new Report.Failure.Builder()
                             .message(failure.file().getName() + ":" + failure.line() + ": " + failure.message())
                             .details(
-                                    failure.severity() + ": " + failure.message() + "\n"
-                                            + "Category: " + failure.source() + "\n"
+                                    failure.severity() + ": " + failure.message() + failure.details() + "\n"
+                                            + (failure.source().isEmpty() ? "" : "Category: " + failure.source() + "\n")
                                             + "File: " + rootDir.toPath().relativize(failure.file().toPath()) + "\n"
                                             + "Line: " + failure.line() + "\n")
                             .build())

--- a/src/main/java/com/palantir/gradle/circlestyle/JavacFailuresSupplier.java
+++ b/src/main/java/com/palantir/gradle/circlestyle/JavacFailuresSupplier.java
@@ -1,0 +1,73 @@
+package com.palantir.gradle.circlestyle;
+
+import static java.lang.Integer.parseInt;
+
+import java.io.File;
+import java.io.IOException;
+import java.util.List;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+
+import org.gradle.api.logging.StandardOutputListener;
+import org.gradle.api.tasks.compile.JavaCompile;
+import org.inferred.freebuilder.shaded.com.google.common.collect.ImmutableList;
+
+import com.google.common.annotations.VisibleForTesting;
+
+class JavacFailuresSupplier implements FailuresSupplier {
+
+    public static JavacFailuresSupplier create(final JavaCompile javac) {
+        // Capture standard output
+        final StringBuilder errorStream = new StringBuilder();
+        StandardOutputListener listener = new StandardOutputListener() {
+            @Override
+            public void onOutput(CharSequence output) {
+                errorStream.append(output);
+            }
+        };
+        javac.getLogging().addStandardErrorListener(listener);
+
+        // Configure the finalizer task
+        return new JavacFailuresSupplier(errorStream);
+    }
+
+    private static final Pattern ERROR_LINE = Pattern.compile("([^ ].*):(\\d+): error: (.*)");
+
+    private final StringBuilder errorStream;
+
+    @VisibleForTesting JavacFailuresSupplier(StringBuilder errorStream) {
+        this.errorStream = errorStream;
+    }
+
+    @Override
+    public List<Failure> getFailures() throws IOException {
+        ImmutableList.Builder<Failure> failures = ImmutableList.builder();
+        Failure.Builder failureBuilder = null;
+        StringBuilder details = null;
+        for (String line : errorStream.toString().split("\n")) {
+            if (failureBuilder != null) {
+                if (line.startsWith(" ")) {
+                    details.append("\n").append(line);
+                    continue;
+                } else {
+                    failures.add(failureBuilder.details(details.toString()).build());
+                    failureBuilder = null;
+                    details = null;
+                }
+            }
+            Matcher matcher = ERROR_LINE.matcher(line);
+            if (matcher.matches()) {
+                failureBuilder = new Failure.Builder()
+                        .file(new File(matcher.group(1)))
+                        .line(parseInt(matcher.group(2)))
+                        .severity("ERROR")
+                        .message(matcher.group(3));
+                details = new StringBuilder();
+            }
+        }
+        if (failureBuilder != null) {
+            failures.add(failureBuilder.details(details.toString()).build());
+        }
+        return failures.build();
+    }
+}

--- a/src/main/java/com/palantir/gradle/circlestyle/StyleTaskTimer.java
+++ b/src/main/java/com/palantir/gradle/circlestyle/StyleTaskTimer.java
@@ -23,6 +23,7 @@ import org.gradle.api.execution.TaskExecutionListener;
 import org.gradle.api.plugins.quality.Checkstyle;
 import org.gradle.api.plugins.quality.FindBugs;
 import org.gradle.api.tasks.TaskState;
+import org.gradle.api.tasks.compile.JavaCompile;
 
 class StyleTaskTimer implements TaskExecutionListener {
 
@@ -55,6 +56,6 @@ class StyleTaskTimer implements TaskExecutionListener {
     }
 
     public static boolean isStyleTask(Task task) {
-        return task instanceof Checkstyle || task instanceof FindBugs;
+        return task instanceof Checkstyle || task instanceof FindBugs || task instanceof JavaCompile;
     }
 }

--- a/src/test/java/com/palantir/gradle/circlestyle/CircleStylePluginTests.java
+++ b/src/test/java/com/palantir/gradle/circlestyle/CircleStylePluginTests.java
@@ -58,6 +58,33 @@ public class CircleStylePluginTests {
     }
 
     @Test
+    public void javacIntegrationTest() throws IOException {
+        copyTestFile("non-compiling-class", projectDir, "src/main/java/com/example/MyClass.java");
+
+        BuildResult result = GradleRunner.create()
+                .withProjectDir(projectDir.getRoot())
+                .withArguments("--stacktrace", "compileJava")
+                .buildAndFail();
+        assertThat(result.getOutput())
+                .contains("Compilation failed")
+                .contains("error: incompatible types")
+                .contains("private final int a")
+                .contains("error: cannot assign a value to final variable b")
+                .contains("b = 2")
+                .contains("uses unchecked or unsafe operations");
+
+        File report = new File(reportsDir, "javac/foobar-compileJava.xml");
+        assertThat(report).exists();
+        String reportXml = Files.asCharSource(report, UTF_8).read();
+        assertThat(reportXml)
+                .contains("incompatible types")
+                .contains("private final int a")
+                .contains("cannot assign a value to final variable b")
+                .contains("b = 2")
+                .doesNotContain("uses unchecked or unsafe operations");
+    }
+
+    @Test
     public void checkstyleIntegrationTest() throws IOException {
         copyTestFile("checkstyle-violating-class", projectDir, "src/main/java/com/example/MyClass.java");
 

--- a/src/test/java/com/palantir/gradle/circlestyle/FailuresReportGeneratorTests.java
+++ b/src/test/java/com/palantir/gradle/circlestyle/FailuresReportGeneratorTests.java
@@ -22,6 +22,11 @@ import static com.palantir.gradle.circlestyle.TestCommon.REPORT;
 import static com.palantir.gradle.circlestyle.TestCommon.ROOT;
 import static org.assertj.core.api.Assertions.assertThat;
 
+import java.io.File;
+import java.util.List;
+
+import javax.xml.transform.TransformerException;
+
 import org.junit.Test;
 
 import com.google.common.collect.ImmutableList;
@@ -44,6 +49,55 @@ public class FailuresReportGeneratorTests {
         Report report = failuresReport(
                 ROOT, "fooproject", "checkstyleTest", FAILED_CHECKSTYLE_TIME_NANOS, CHECKSTYLE_FAILURES);
         assertThat(report).isEqualTo(REPORT);
+    }
+
+    @Test
+    public void testJavacErrors() throws TransformerException {
+        List<Failure> failures = ImmutableList.of(
+                new Failure.Builder()
+                        .file(new File(ROOT, "src/main/java/com/example/MyClass.java"))
+                        .line(8)
+                        .severity("ERROR")
+                        .message("incompatible types: String cannot be converted to int")
+                        .details("\n    private final int a = \"hello\";                               "
+                                + "\n                          ^")
+                        .build(),
+                new Failure.Builder()
+                        .file(new File(ROOT, "src/main/java/com/example/MyClass.java"))
+                        .line(12)
+                        .severity("ERROR")
+                        .message("cannot assign a value to final variable b")
+                        .details("\n        b = 2;                                                   "
+                                        + "\n        ^                                                        ")
+                        .build());
+        Report report = failuresReport(ROOT, "foobar", "compileJava", 293_000, failures);
+        assertThat(report).isEqualTo(new Report.Builder()
+                .name("foobar")
+                .subname("compileJava")
+                .elapsedTimeNanos(293_000)
+                .addTestCases(new Report.TestCase.Builder()
+                        .name("com.example.MyClass")
+                        .failure(new Report.Failure.Builder()
+                                .message("MyClass.java:8: incompatible types: String cannot be converted to int")
+                                .details("ERROR: incompatible types: String cannot be converted to int\n"
+                                        + "    private final int a = \"hello\";                               \n"
+                                        + "                          ^\n"
+                                        + "File: src/main/java/com/example/MyClass.java\n"
+                                        + "Line: 8\n")
+                                .build())
+                        .build())
+                .addTestCases(new Report.TestCase.Builder()
+                        .name("com.example.MyClass")
+                        .failure(new Report.Failure.Builder()
+                                .message("MyClass.java:12: cannot assign a value to final variable b")
+                                .details("ERROR: cannot assign a value to final variable b\n"
+                                        + "        b = 2;                                                   \n"
+                                        + "        ^                                                        \n"
+                                        + "File: src/main/java/com/example/MyClass.java\n"
+                                        + "Line: 12\n")
+                                .build())
+                        .build())
+                .build());
 
     }
 }

--- a/src/test/java/com/palantir/gradle/circlestyle/JavacFailuresSupplierTest.java
+++ b/src/test/java/com/palantir/gradle/circlestyle/JavacFailuresSupplierTest.java
@@ -1,0 +1,60 @@
+package com.palantir.gradle.circlestyle;
+
+import static org.assertj.core.api.Assertions.assertThat;
+
+import java.io.File;
+import java.io.IOException;
+
+import org.junit.Test;
+
+public class JavacFailuresSupplierTest {
+
+    private static final String CLASS_FILE = "/tmp/ab1/src/main/java/com/example/MyClass.java";
+    private static final int LINE_1 = 8;
+    private static final String ERROR_1 = "incompatible types: String cannot be converted to int";
+    private static final String DETAIL_1 = "\n    private final int a = \"hello\";                               "
+            + "\n                          ^";
+    private static final int LINE_2 = 12;
+    private static final String ERROR_2 = "cannot assign a value to final variable b";
+    private static final String DETAIL_2 = "\n        b = 2;                                                   "
+                    + "\n        ^                                                        ";
+
+    @Test
+    public void noFailuresInEmptyOutput() throws IOException {
+        String javacOutput = "";
+        JavacFailuresSupplier supplier = new JavacFailuresSupplier(new StringBuilder(javacOutput));
+        assertThat(supplier.getFailures()).isEmpty();
+    }
+
+    @Test
+    public void noFailuresInOutputWithOnlyWarnings() throws IOException {
+        String javacOutput = "warning: [options] bootstrap class path not set in conjunction with -source 1.7\n"
+                + "Note: " + CLASS_FILE + " uses unchecked or unsafe operations.\n"
+                + "Note: Recompile with -Xlint:unchecked for details.               \n"
+                + "1 warning";
+        JavacFailuresSupplier supplier = new JavacFailuresSupplier(new StringBuilder(javacOutput));
+        assertThat(supplier.getFailures()).isEmpty();
+    }
+
+    @Test
+    public void twoFailuresInOutputWithNoWarnings() throws IOException {
+        String javacOutput = CLASS_FILE + ":" + LINE_1 + ": error: " + ERROR_1 + DETAIL_1 + "\n"
+                + CLASS_FILE + ":" + LINE_2 + ": error: " + ERROR_2 + DETAIL_2 + "\n";
+        JavacFailuresSupplier supplier = new JavacFailuresSupplier(new StringBuilder(javacOutput));
+        assertThat(supplier.getFailures()).containsExactly(
+                new Failure.Builder()
+                        .file(new File(CLASS_FILE))
+                        .line(LINE_1)
+                        .severity("ERROR")
+                        .message(ERROR_1)
+                        .details(DETAIL_1)
+                        .build(),
+                new Failure.Builder()
+                        .file(new File(CLASS_FILE))
+                        .line(12)
+                        .severity("ERROR")
+                        .message(ERROR_2)
+                        .details(DETAIL_2)
+                        .build());
+    }
+}

--- a/src/test/resources/com/palantir/gradle/circlestyle/non-compiling-class
+++ b/src/test/resources/com/palantir/gradle/circlestyle/non-compiling-class
@@ -1,0 +1,15 @@
+package com.example;
+
+import java.util.ArrayList;
+import java.util.List;
+
+public class MyClass {
+
+    private final int a = "hello";
+    private final int b = 1;
+    
+    public void methodA() {
+        b = 2;
+        List<Integer> foo = new ArrayList();
+    }
+}

